src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs

```diff
@@ -35,20 +36,59 @@ public override string Run(TaskContext context, string input)
 
         public async override Task<string> RunAsync(TaskContext context, string rawInput)
         {
-            HttpRequestMessage requestMessage = await this.ReconstructHttpRequestMessage(rawInput);
-            HttpResponseMessage response = await this.httpClient.SendAsync(requestMessage);
-            DurableHttpResponse durableHttpResponse = await DurableHttpResponse.CreateDurableHttpResponseWithHttpResponseMessage(response);
+            DurableHttpRequest durableHttpRequest = ReconstructDurableHttpRequest(rawInput);
+            HttpRequestMessage requestMessage = await this.ReconstructHttpRequestMessage(durableHttpRequest);
 
-            return JsonConvert.SerializeObject(durableHttpResponse);
+            CancellationTokenSource cts = new CancellationTokenSource();
+            if (durableHttpRequest.Timeout != null)
+            {
+                try
+                {
+                    TimeSpan timeout = durableHttpRequest.TimeoutExpirationDateTime - DateTime.UtcNow;
+                    this.httpClient.Timeout = timeout;
+                    cts.CancelAfter(timeout);
+                }
+                catch (Exception)
+                {
+                    TimeSpan timeout = durableHttpRequest.Timeout.Value;
+                    this.httpClient.Timeout = timeout;
+                    cts.CancelAfter(timeout);
+                }
+            }
+
+            try
+            {
+                HttpResponseMessage response;
+                if (durableHttpRequest.Timeout == null)
+                {
+                    response = await this.httpClient.SendAsync(requestMessage);
+                }
+                else
+                {
+                    response = await this.httpClient.SendAsync(requestMessage, cts.Token);
+                }
+
+                DurableHttpResponse durableHttpResponse = await DurableHttpResponse.CreateDurableHttpResponseWithHttpResponseMessage(response);
+
+                return JsonConvert.SerializeObject(durableHttpResponse);
+            }
+            catch (OperationCanceledException ex) // when (cts.Token.IsCancellationRequested)
```

Comment:
To make this more accurate, it could be `catch (OperationCanceledException ex) when (cts.Token.IsCancellationRequested)` (added the commented out part). This way we know that the exception came when a cancel request was sent to `CancellationTokenSource`. The only issue with this is adding tests that mirror the action of cancelling a `CancellationTokenSource`. The `Mock<HttpMessageHandler>` Setup call takes `ItExpr.IsAny<CancellationToken>()` and the test would need the `CancellationTokenSource` to cancel the task, not the `CancellationToken`.

I saw that that `ItExpr.IsAny()` could evaluate conditions, but didn't see anything about sending in another object to manipulate. Let me know if there's a way to pass in a CancellationTokenSource using Moq.

This is how the `HttpMessageHandler` is mocked right now:
```
var handlerMock = new Mock<HttpMessageHandler>(MockBehavior.Strict);
handlerMock 
  .Protected()
  .Setup<Task<HttpResponseMessage>>("SendAsync", ItExpr.IsAny<HttpRequestMessage>(), ItExpr.IsAny<CancellationToken>())
  .ReturnsAsync(httpResponseMessage);

```